File: android/app/src/test/java/com/dbchbin/ompgui/remote/relay/PairingUriTest.kt
package com.dbchbin.ompgui.remote.relay

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test
import java.net.URLEncoder
import java.nio.charset.StandardCharsets

class PairingUriTest {
    @Test
    fun parsesFragmentPairingUri() {
        val url = URLEncoder.encode("wss://mac.example.ts.net/relay", StandardCharsets.UTF_8.name())
        val uri = "ompgui://pair#v=1&url=$url&sid=s_abcDEF1234567890&secret=sekritvalue_0123456789abcdefghijk"
        val parsed = parsePairingUri(uri)
        assertEquals(
            PairingOffer(
                url = "wss://mac.example.ts.net/relay",
                serverId = "s_abcDEF1234567890",
                secret = "sekritvalue_0123456789abcdefghijk",
            ),
            parsed,
        )
    }

    @Test
    fun rejectsForeignSchemesAndQuerySecrets() {
        assertNull(
            parsePairingUri(
                "https://pair#v=1&url=wss://mac.example.ts.net/relay&sid=s_abcDEF1234567890&secret=sekritvalue_0123456789abcdefghijk",
            ),
        )
        assertNull(
            parsePairingUri(
                "ompgui://pair?v=1&url=wss://mac.example.ts.net/relay&sid=s_abcDEF1234567890&secret=sekritvalue_0123456789abcdefghijk",
            ),
        )
    }

    @Test
    fun scannerAcceptsValidPairCodeWithWhitespace() {
        // QR payloads may carry stray whitespace; the scanner gates on this parse.
        val url = URLEncoder.encode("wss://mac.example.ts.net/relay", StandardCharsets.UTF_8.name())
        val uri = "  ompgui://pair#v=1&url=$url&sid=s_abc&secret=topsecret123  "
        val parsed = parsePairingUri(uri)
        assertEquals(
            PairingOffer(
                url = "wss://mac.example.ts.net/relay",
                serverId = "s_abc",
                secret = "topsecret123",
            ),
            parsed,
        )
    }

    @Test
    fun scannerRejectsNonPairingQrPayloads() {
        assertNull(parsePairingUri("https://example.com/not-a-pairing-code"))
        assertNull(parsePairingUri("ompgui://pair#v=1&url=&sid=&secret="))
        assertNull(
            parsePairingUri(
                "ompgui://pair#v=1&url=wss://mac.example.ts.net/relay&sid=bad sid&secret=topsecret123",
            ),
        )
    }
}

File: lib/relay/pairing-uri.test.mjs
import assert from "node:assert/strict";
import test from "node:test";
import { createJiti } from "jiti";

const jiti = createJiti(import.meta.url, { tsconfigPaths: true });
const { buildPairingUri } = await jiti.import("./pairing-uri.ts");

test("pairing offer keeps encoded endpoint and credentials in its fragment", () => {
  const uri = new URL(buildPairingUri({
    version: 1,
    url: "wss://mac.example.ts.net/relay?key=a&other=b",
    serverId: "s_abcDEF1234567890",
    secret: "sekritvalue_0123456789abcdefghijk",
  }));
  assert.equal(uri.protocol, "ompgui:");
  assert.equal(uri.host, "pair");
  assert.equal(uri.search, "");
  assert.deepEqual(Object.fromEntries(new URLSearchParams(uri.hash.slice(1))), {
    v: "1", url: "wss://mac.example.ts.net/relay?key=a&other=b",
    sid: "s_abcDEF1234567890", secret: "sekritvalue_0123456789abcdefghijk",
  });
});
